[Feature] Integrate PDI Interface enhancements & discuss directory placement

mfahdaz opened the pull request:
**Overview**

This pull request integrates PDI interface enhancements, originally developed in `parflow-pdi` (a GitLab fork of Parflow), into the latest Parflow codebase. The implementation follows [PDI](https://pdi.dev/1.8/index.html) guidelines and introduces new functionality to support in-situ data processing and analysis.

**What is PDI?**

For more details, please refer to the [PDI GitHub repository](https://github.com/pdidev/pdi/tree/main?tab=readme-ov-file#the-pdi-distribution), where it is described as:

> "PDI is a library that aims to decouple high-performance simulation codes from input/output concerns. It offers a declarative API that enables codes to expose the buffers in which they store data and to notify PDI of significant simulation steps. Additionally, it supports a plugin system to make existing libraries such as HDF5, NetCDF, or Python available to codes—potentially mixing several in a single execution."

**Implemented Features**

Current iteration implemented the following PDI features:

- Per-Process Data Collection: Each process collects simulation data independently.
- HDF5 File Output: Data is saved to HDF5 files with filenames that include the corresponding process rank.

**Future Enhancements**

- In-Transit Data Processing: Future iterations will focus on in-transit data processing and analysis, reducing the need for disk I/O operations.

**What This PR Does**

This PR lays the foundation for integrating PDI into Parflow, and aims to initiate discussions around the current implementation and future development of in-situ data processing capabilities.

- Upstream Sync: Merges the latest changes from `parflow-pdi`, ensuring compatibility with current upstream developments.
- PDI Interface Integration: Adds new components and modifications necessary for PDI support.
- Conflict Resolution & Refactoring: Updates existing modules to address conflicts and align with upstream code changes.
- Temporary Directory Structure: Configuration file for PDI specification tree (defining data types and plugin lists) and output testing files are currently located in a dedicated `pdi/` directory.

**Discussion Points**

We would appreciate feedback on the following:

- Directory Placement: What would be the best location or approach to organize all PDI-related functionality within Parflow, considering its evolved structure? For example, would placing it under `pfsimulator/third_party/` be appropriate, or might there be a more suitable alternative?
- Naming Conventions & Architecture: Are there specific naming conventions or architectural guidelines that should be followed to better align with Parflow’s structure?

We welcome feedback on the best way to integrate these changes, both in terms of functionality and repository structure.

Thank you!

smithsg84 reviewed (changes requested):
Thanks for the PDI addition.

I do think having the extras in third_party is a better place than at the root of the tree.

I'm thinking the Python code should go into the pftools so it would be installed with the python pftools library.   The pftools/python/parflow has cli directory for command line tools and library code is in tools.   There is already a compare.py method available there.

Documentation needs to be added for the new keys, "keys.rst" file and solver.yaml for the pfkeys system.

smithsg84 commented:
How easy is it to add a test to our CI system?   Are the PDI libraries easy/quick to build?

mfahdaz commented on cmake/modules/FindPDI.cmake:
```diff
@@ -0,0 +1,54 @@
+MARK_AS_ADVANCED(PARACONF_INCLUDE_DIRS PARACONF_LIBRARIES)
+
+
+# Find SPDLOG library (libspdlog.so)
```
Created separate CMake modules for SPDLOG and PARACONF, independent of PDI, allowing users to specify `<module_name>_ROOT` (or set the environment variable `<module_name>_ROOT`) to guide the search for modules in non-standard locations. If neither is specified and `PDI_ROOT` is defined, the search will default to `${PDI_ROOT}`

mfahdaz commented on pfsimulator/parflow_lib/write_parflow_binary.c:
```diff
@@ -109,12 +109,20 @@ void       WritePFBinary_Subvector(
    *  this is way to slow we need better boxloops that operate
    *  on arrays */
 
+  // printf("nx_v: %d %d %d\n",nx_v,ny_v,nz_v);
```
Done!

mfahdaz commented on pfsimulator/parflow_lib/write_parflow_pdi.c:
```diff
@@ -0,0 +1,150 @@
+#endif
+
+
+void     WritePDI(
```
Done!

mfahdaz commented on pfsimulator/parflow_lib/write_parflow_pdi.c:
```diff
@@ -0,0 +1,150 @@
+  char filename[255] = "";
+  sprintf(filename, "%s.%s.h5.%d", file_prefix, file_suffix, p);
+  
+  // if (p == 0)
```
Done!

mfahdaz commented on pfsimulator/parflow_lib/write_parflow_pdi.c:
```diff
@@ -0,0 +1,150 @@
+  long size;
+  
+  // Same timers as for PFB files
+  BeginTiming(PFBTimingIndex);
```
Done!

mfahdaz commented:
> Thanks for the PDI addition.
> 
> I do think having the extras in third_party is a better place than at the root of the tree.
> 
> I'm thinking the Python code should go into the pftools so it would be installed with the python pftools library. The pftools/python/parflow has cli directory for command line tools and library code is in tools. There is already a compare.py method available there.
> 
> Documentation needs to be added for the new keys, "keys.rst" file and solver.yaml for the pfkeys system.

Thank you for your valuable suggestions! Your feedback has greatly helped improve the clarity and structure of these changes.

**Summary of Major Changes in PR**:

- Moved PDI to the `third_party` directory.
- Relocated Python scripts for comparing PDI output with PF Binary to `pftools/python/parflow/cli`.
- Refactored code for improved readability, maintainability, and efficiency.
- Added a new timer for PDI.
- Introduced standalone CMake modules for SPDLOG and PARACONF, independent of PDI, allowing users to specify `<module_name>_ROOT` (or set the corresponding environment variable) to guide module searches in non-standard locations. If unspecified, but `PDI_ROOT` is defined, the search defaults to `${PDI_ROOT}`.
- Added a CI test for ParFlow build and installation with PDI.
- Added `README-PDI.md` with an overview of PDI, installation guide, and related details.
- Updated documentation to include details on newly introduced keys.

mfahdaz commented on CMakeLists.txt:
```diff
@@ -299,6 +299,21 @@ if (${PARFLOW_ENABLE_HDF5} OR DEFINED HDF5_ROOT)
+  # These libraries must be found to ensure proper functionality.
+  find_package(PARACONF REQUIRED)
+  find_package(SPDLOG REQUIRED)
+endif (${PDI_FOUND})
```
Done!

mfahdaz commented on README-PDI.md:
````diff
@@ -0,0 +1,178 @@
+
+```shell
+scp /parflow/pfsimulator/third_party/pdi/conf.yml .
+```
````
Done!

mfahdaz commented on pf-keys/definitions/solver.yaml:
```diff
@@ -2634,6 +2638,174 @@ Solver:
+  
+  WritePDISaturation:
+    help: >
+      [Type: boolean/string] This key is used to specify exposre of the saturation
```
Done!

mfahdaz commented on pf-keys/definitions/solver.yaml:
```diff
@@ -2634,6 +2638,174 @@ Solver:
       BoolDomain:
       RequiresModule: SILO
 
+  # PDI outputs
```
Thanks for pointing that out! In both places: `docs/user_manual/keys.rst` and this file (`solver.yaml`), the PDI documentation is part of the Solver's output options section and follows the "SILO outputs" entry. So the order is consistent in that context.

mfahdaz commented on pfsimulator/third_party/pdi/README.md:
````diff
@@ -0,0 +1,107 @@
+
+```
+python $PARFLOW_DIR/pftools/python/parflow/cli/compare_pdi_pfb.py <file name without extension>
+```
````
This content was initially included for completeness. I've replaced the repeated content with a reference to `README-PDI.md` to avoid redundancy and keep the documentation clear and centralized.

mfahdaz commented on pftools/python/parflow/cli/compare_pdi_pfb.py:
```diff
@@ -0,0 +1,345 @@
+            np.abs(pfb_subvectors[igrid]["data"] - pdi_subvectors[igrid]["data"])
+        )
+
+    print(f" - Error: {error}")
```
This script is currently designed as a self-sufficient cli tool to compare the PDI-generated (HDF5) files with PFB binary files for output verification. In future iterations, we plan to integrate in-situ functionality and refactor it to reuse the functionality from `compare.py`.

smithsg84 reviewed (approved):
Thanks for update and addressing comments.
